src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +550,123 @@ ApplicationImpl::joinAllThreads()
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Manually closed ledger "
+                                       "sequence number already at max ({})"),
+                            startLedgerSeq));
+        }
+        auto minLedgerSeq = startLedgerSeq + 1;
+        auto const targetLedgerSeq =
+            manualLedgerSeq ? *manualLedgerSeq : minLedgerSeq;
+        if (targetLedgerSeq < minLedgerSeq)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Invalid manual close ledger "
+                                       "sequence number {} (minimum {})"),
+                            targetLedgerSeq, minLedgerSeq));
+        }
+        auto const lastCloseTime = getLedgerManager()
+                                       .getLastClosedLedgerHeader()
+                                       .header.scpValue.closeTime;
+        auto const now = VirtualClock::to_time_t(getClock().system_now());
+        uint32_t const defaultManualCloseTimeInterval = 5;
+        auto nextCloseTime = lastCloseTime + 1;
+        auto const defaultNextCloseTime =
+            lastCloseTime + defaultManualCloseTimeInterval;
+        if (nextCloseTime < now)
+        {
+            nextCloseTime = now;
+        }
+        if (manualCloseTime)
+        {
+            if (*manualCloseTime < nextCloseTime)
+            {
+                throw std::invalid_argument(
+                    fmt::format(FMT_STRING("Invalid manual close time {} (last "
+                                           "close time={}, now={})"),
+                                *manualCloseTime, lastCloseTime, now));
+            }
+            nextCloseTime = *manualCloseTime;
+        }
+        else
+        {
+            if (nextCloseTime < defaultNextCloseTime)
+            {
+                nextCloseTime = defaultNextCloseTime;
+            }
+        }
+        if (targetLedgerSeq != minLedgerSeq)
+        {
+            if (!mHerder->isTransactionQueueEmpty())
+            {
+                throw std::invalid_argument(
+                    fmt::format(FMT_STRING("May not skip ledgers (requested "
+                                           "sequence number {}; next {}) while "
+                                           "transaction queue is non-empty"),
+                                targetLedgerSeq, minLedgerSeq));
+            }
+            LOG(INFO) << "closing ledger with sequence number "
+                      << targetLedgerSeq - 1 << " on behalf of target "
+                      << targetLedgerSeq;
+            LedgerTxn ltx(getLedgerTxnRoot());
+            ltx.loadHeader().current().ledgerSeq = targetLedgerSeq - 1;
+            getLedgerManager().manuallyAdvanceLedgerHeader(
+                ltx.loadHeader().current());
+            ltx.commit();
+            auto const& header =
+                getLedgerManager().getLastClosedLedgerHeader().header;
+            dynamic_cast<HerderImpl&>(getHerder())
```

Comment:
yes we need to do this.
I'd suggest adding a new method to Herder to expose this in a slightly more sane way and with a name that yells "unsafe".

Maybe something like `forceSyncedState()` (no parameters) that calls `restoreSCPState` with data gathered from lcl.

Reply:
I've given it the descriptive-but-hostile-sounding name `forceSCPStateIntoSyncWithLastClosedLedger()`, though I'll shorten it to `forceSyncedState()` (or whatever) if you prefer.